File: src/java/com/codecool/gameoflife/view/GameView.java
package com.codecool.gameoflife.view;

import com.codecool.gameoflife.controller.GameController;
import com.codecool.gameoflife.model.Cell;
import javafx.scene.layout.HBox;
public class GameView extends HBox {

    private GridPaneView gameBoardView;
    private SetupView setupView = new SetupView();
    private GameController gameController;
    private GameLoop gameAnimation;

    public GameView(GameController gameController) {
        this.gameBoardView = new GridPaneView(gameController);
        this.gameController = gameController;
        this.gameAnimation = new GameLoop(200, gameController::nextGeneration);
        this.setupView.setupStartButton(event -> this.startGame(), "Start Game");
        this.setupView.setupRandomizeButton(event -> this.gameController.randomizeGrid());
        this.setupView.setupClearButton(event -> this.gameController.clearBoard());
        this.setupView.setupSlider((observable, oldValue, newValue) -> this.changeGameSpeed(newValue.intValue()));
        this.setupView.setupBorderLessToggle((observable, oldValue, newValue) -> this.gameController.toggleBorderlessMode());
        this.setupView.setupGridVisibleToggle((observable, oldValue, newValue) -> this.setGridVisible(newValue));
        this.setupView.createRuleButtons(gameController);
        this.getChildren().addAll(gameBoardView, setupView);
    }

    private void changeGameSpeed(int gameDelay) {
        this.gameAnimation.setDelay(gameDelay);
    }

    private void setGridVisible(boolean visible) {
        this.gameBoardView.setGridOn(visible);
        this.updateView(this.gameController.getGameBoard());
    }

    public void updateView(Cell[][] gameBoard) {
        this.gameBoardView.updateCells(gameBoard);
    }

    private void startGame() {
        this.gameAnimation.start();
        this.setupView.setupStartButton(event -> this.stopGame(), "Stop");
    }

    private void stopGame() {
        this.gameAnimation.stop();
        this.setupView.setupStartButton(event -> this.startGame(), "Start Game");
    }

    public void updateCell(Cell cellToUpdate, int x, int y) {
        this.gameBoardView.updateSingleCell(cellToUpdate, x, y);
    }

    public void updateRules() {
        this.setupView.createRuleButtons(this.gameController);
    }
}

File: src/kotlin/com/codecool/gameoflife/view/GameView.kt
package com.codecool.gameoflife.view

import com.codecool.gameoflife.controller.GameController
import com.codecool.gameoflife.model.Cell
import javafx.beans.value.ChangeListener
import javafx.event.EventHandler
import javafx.scene.layout.HBox

class GameView(private val gameController: GameController) : HBox() {

    private val gameBoardView: GridPaneView = GridPaneView(gameController)
    private val gameAnimation: GameLoop = GameLoop(200, Runnable { gameController.nextGeneration() })
    private val setupView: SetupView

    init {
        setupView = SetupView().apply {
            setupStartButton(EventHandler { startGame() }, "Start Game")
            setupRandomizeButton(EventHandler { gameController.randomizeGrid() })
            setupClearButton(EventHandler { gameController.clearBoard() })
            setupSlider(ChangeListener { _, _, newValue -> changeGameSpeed(newValue.toInt()) })
            setupBorderLessToggle(ChangeListener { _, _, _ -> gameController.toggleBorderlessMode() })
            setupGridVisibleToggle(ChangeListener { _, _, newValue -> setGridVisible(newValue) })
            createRuleButtons(gameController)
        }
        children.addAll(gameBoardView, setupView)
    }

    private fun changeGameSpeed(gameDelay: Int) {
        gameAnimation.setDelay(gameDelay)
    }

    private fun setGridVisible(visible: Boolean) {
        gameBoardView.gridOn = visible
        updateView(gameController.gameBoard.board)
    }

    fun updateView(gameBoard: Array<Array<Cell>>) {
        gameBoardView.updateCells(gameBoard)
    }

    private fun startGame() {
        gameAnimation.start()
        setupView.setupStartButton(EventHandler { stopGame() }, "Stop")
    }

    private fun stopGame() {
        gameAnimation.stop()
        setupView.setupStartButton(EventHandler { startGame() }, "Start Game")
    }

    fun updateCell(x: Int, y: Int) {
        gameBoardView.updateSingleCell(x, y)
    }

    fun updateRules() {
        setupView.createRuleButtons(gameController)
    }
}
